lib/sensors/src/URM04Sensor.cpp

```diff
@@ -1,197 +1,124 @@
+  cmdst[5] = checkSum;
+
+  /******** SEND COMMANDS OVER SERIAL *********************/
+  int num_bytes;
+  // returns the number of bytes if successful write
+  num_bytes = serial.write(&cmdst[0], sizeof(cmdst));
+
+  // else if numb_btyes < 0 serial failed or if number of bytes written != 6 (because 6 bytes sent over serial)
+  if (num_bytes < 0 || num_bytes != 6) {
+    success = false;
+  }
+
+  // flush the buffer from serial
+  serial.sync();
+
+  // wait for at least 30 ms after calling trigger function
+  ThisThread::sleep_for(std::chrono::milliseconds(35));
+
+  // reset command buffer - fill whole array with zeros
+  memset(&cmdst[0], 0, sizeof(cmdst));
 }
-// reads the distance from URM04
-void sensor::URM04Sensor::urmReader(int id) {
-  // fill command buffer with read commands
-  cmdst[2] = id;
+
+float URM04Sensor::URM04Sensor::read_distance() {
+  /************ INSTANTIATE COMMANDS TO BE SENT OVER SERIAL************/
+  // check sum represents the final bit in command buffer - made by adding all previous bits in command buffer
+  uint8_t checkSum;
+  // buffer header
+  cmdst[0] = 0x55;
+  cmdst[1] = 0xAA;
+  // device address
+  cmdst[2] = startAddr;
+  // command length
   cmdst[3] = 0x00;
+  // the command itself
   cmdst[4] = 0x02;
-  // send command over serial
-  transmitCommands();
-}
-// transmit commands
-void sensor::URM04Sensor::transmitCommands() {
-  cmdst[5] = cmdst[0] + cmdst[2] + cmdst[3] + cmdst[4];
-  // delay for 1 second using thread  1000ms = 1sec
-  ThisThread::sleep_for(std::chrono::milliseconds(1000));
-  // send command over serial connection with BufferedSerial write and put into cmdst
-  nucleo_board.write(&cmdst[0], sizeof(cmdst));
-  // delay for 2 seconds 2000ms = 2sec
-  ThisThread::sleep_for(std::chrono::milliseconds(2000));
-}
-
-// this function actually gets the data
-void sensor::URM04Sensor::analyzeUrmData(uint8_t cmd[]) {
-  uint8_t checkSum = 0;
-  // add each byte in the command array to create checksum
-  for (int i{0}; i < 7; i++) {
-    checkSum += cmd[i];
-  }
-  // check is checksum is correct and that the commands are correct as well
-  if (checkSum == cmd[7] && cmd[3] == 2 && cmd[4] == 2) {
-    uint8_t id  = cmd[2] - startAddr;
-    urmData[id] = cmd[5] * 256 + cmd[6];
-  }
-  // else an error occured during transmission
-  else if (cmd[3] == 2 && cmd[4] == 2) {
-    read_success = false;
+  // checksum bit
+  checkSum = cmdst[0] + cmdst[1] + cmdst[2] + cmdst[3] + cmdst[4];
+  // instantiate the last element in the command buffer with checksum value
+  cmdst[5] = checkSum;
+
+  /************** SEND COMMANDS OVER SERIAL***********/
+  int w_num_bytes;
+  // returns the number of bytes if successful read
+  w_num_bytes = serial.write(&cmdst[0], sizeof(cmdst));
+
+  // else if numb_btyes < 0 serial failed or if number of bytes read != 6 (because 6 bytes sent over serial)
+  if (w_num_bytes < 0 || w_num_bytes != 6) {
+    success = false;
```

Comment:
dont take this approach with success = false. if the function fails return false. and the value that is read should be put in the argument which u take in by ref